File: Data Structures & Algorithms/binary-tree-diameter/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
     int maxi = 0;
    int diameterOfBinaryTree(TreeNode* root) {
        if(!root) return 0;
        auto left = dfs(root->left);
        auto right = dfs(root->right);
        maxi = max(maxi, left+right);
        diameterOfBinaryTree(root->left);
        diameterOfBinaryTree(root->right);
        return maxi;
    }
    int dfs(TreeNode * root){
        if(!root) return 0;
        return 1+max(dfs(root->left), dfs(root->right));
    }
};

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        self.pdep = -1
        if not root:
            return []
        res = []
        def dfs(root, dep):
            if not root:
                return []
            if dep > self.pdep:
                res.append(root.val)
                self.pdep = dep

            dfs(root.right, dep+1)
            dfs(root.left, dep+1)
        dfs(root, 0)
        return res